rust-runtime/aws-smithy-http-server/src/routing/routers/mod.rs

```diff
@@ -0,0 +1,144 @@
+use std::{
+    fmt::Debug,
+    future::{ready, Future, Ready},
+    marker::PhantomData,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use futures_util::future::Either;
+use http::Response;
+use tower::{util::Oneshot, Service, ServiceExt};
+
+pub mod aws_json;
+// pub mod merged;
+pub mod rest;
+
+use crate::{
+    body::{empty, BoxBody},
+    response::IntoResponse,
+};
+
+const UNKNOWN_OPERATION_EXCEPTION: &str = "UnknownOperationException";
+
+fn method_disallowed() -> http::Response<BoxBody> {
+    http::Response::builder()
+        .status(http::StatusCode::METHOD_NOT_ALLOWED)
+        .body(empty())
+        .expect("valid HTTP response")
+}
+
+pub trait Router<B> {
+    type Service;
+    type Error;
+
+    fn match_route(&self, request: &http::Request<B>) -> Result<Self::Service, Self::Error>;
+}
+
+pub struct RoutingService<R, P> {
+    router: R,
+    _protocol: PhantomData<P>,
+}
+
+impl<R, P> Debug for RoutingService<R, P>
+where
+    R: Debug,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("RoutingService")
+            .field("router", &self.router)
+            .field("_protocol", &self._protocol)
+            .finish()
+    }
+}
+
+impl<R, P> Clone for RoutingService<R, P>
+where
+    R: Clone,
+{
+    fn clone(&self) -> Self {
+        Self {
+            router: self.router.clone(),
+            _protocol: PhantomData,
+        }
+    }
+}
+
+impl<R, P> RoutingService<R, P> {
+    pub fn new(router: R) -> Self {
+        Self {
+            router,
+            _protocol: PhantomData,
+        }
+    }
+
+    pub fn map<RNew, F: FnOnce(R) -> RNew>(self, f: F) -> RoutingService<RNew, P> {
+        RoutingService {
+            router: f(self.router),
+            _protocol: PhantomData,
+        }
+    }
+}
+
+type EitherOneshotReady<S, B> = Either<
```

Comment:
Love how the types and bounds work out beautifully in this file below this line.